### PR TITLE
GPII-3756: Don't destroy env after failure.

### PR DESCRIPTION
**Pull request by mrtyler**
EDIT: Tested this in CI: https://gitlab.com/gpii-ops/gpii-infra/pipelines/49777249. It failed due to unrelated reasons, but the logic looks correct so far (destroyed things first, continuing in spite of failure; ran display_cluster_state when rake failed, then exited).

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -218,19 +219,20 @@ gcp-dev:
...
-    # Clean up even if something failed.
-    - cd gcp/live/dev
-    - rake display_cluster_state
     # We need to chain the following into a single command to prevent various issues
     # caused by subsequent after script commands when destroy task fails.
     # More info: https://issues.gpii.net/browse/GPII-3488
     - (rake destroy && rake destroy_secrets && rake destroy_tfstate[k8s]) || true
     # Terraform state encryption key is also gone with secrets, we have to destroy tfstate as well
     - rake destroy_tfstate[locust] || true
+    - rake || (rake display_cluster_state ; false)
```
Comment (stepanstipl) [inline .gitlab-ci.yml]: Why all the `; false` bits?

Comment (mrtyler) [reply .gitlab-ci.yml]: Without the `false`, when `rake` fails but `rake display_cluster_state` succeeds, the entire line succeeds and the build continues. This is not desirable.

**Comment by mrtyler** [comment]
> So have do we plan to operate - every failed CI run will require manual intervention to clean up?

No, because each run starts by running `rake destroy`.

> Also can't we move some of the `rake display_cluster_state` and/or `detroy_*` tasks to the after script and avoid repetition?

No, because then they will always run, including after an environment is successfully created and then destroyed (when display_cluster_state can not run because there is no cluster for which to display the state).

> Can you describe what the desired high-level behavior is?

Is it unclear in the ticket?

> Ie. stop on any error, leave things as they are and display state?

Yes. Also, destroy everything at the start of the next run.

**Comment by stepanstipl** [comment]
@mrtyler ok cool, thanks for answers, it makes more sense now to me. LGTM

What do you think about splitting up the cleanup info a separate job? Then we would be able to use the `when: on_failure` and `when: on_success` logic of Gitlab pipelines?

What I mean is if you look at our `.gitlab-ci.yml`as a whole, it's a lot of repetition, lot of comments on special cases and lots of constructs like `|| true`, `&& ...`, `; false`. I think it's getting a bit difficult to read, maybe too much logic in there (given we have rake tasks for this). So I'm thinking about how to make it perhaps simpler and better structured.

**Comment by mrtyler** [comment]
> What do you think about splitting up the cleanup info a separate job? Then we would be able to use the when: on_failure and when: on_success logic of Gitlab pipelines?

I think this will add several extra steps to the pipeline. I'm not sure it will make the workflow any clearer. It would give us the ability to retry individual "sub-steps" (only deploy again, only destroy again) though in practice that might mean we have to retry both a destroy step and a deploy step, instead of "retrying gcp-dev".

> What I mean is if you look at our .gitlab-ci.ymlas a whole, it's a lot of repetition, lot of comments on special cases and lots of constructs like || true, && ..., ; false. I think it's getting a bit difficult to read, maybe too much logic in there (given we have rake tasks for this). So I'm thinking about how to make it perhaps simpler and better structured.

I agree that the workflow is hard to understand. I spent a little time thinking about this problem (even before posting this PR) and concluded:
* The workflow *is* fairly complex -- run step X, and only if X succeeds run Y and Z; always run step B, even if step A fails; etc.
* My plan was to get this change out there to start collecting information about numerous pipeline-breaking failures. To this end, I merged this PR as-is.

I started migrating steps into rake, where we have finer-grained control. I'm still not sure if this is worth the effort right now, but if it is I'll post a follow-up PR.

**Comment by mrtyler** [comment]
Did some refactoring in https://github.com/gpii-ops/gpii-infra/pull/320

**Pull request merged**